src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int segmentsCount = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                segmentsCount--;
+                lastSegment = segments[segmentsCount];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, segmentsCount);
 
             // Remove the unnecessary type cast segment.
             ODataPathSegment nextToLastSegment = segments.Last();
+            int lastIndex = segments.Count - 1;
             while (nextToLastSegment is TypeSegment)
             {
                 ODataPathSegment previousSegment = segments[segments.Count - 2];
                 IEdmStructuredType owningType = previousSegment.TargetEdmType as IEdmStructuredType;
                 if (owningType != null && owningType.FindProperty(lastSegment.Identifier) != null)
                 {
-                    segments.Remove(nextToLastSegment);
+                    segments.RemoveAt(lastIndex);
                     nextToLastSegment = segments.Last();
+                    lastIndex = segments.Count - 1;

```

Comment:
@marabooy [This](https://github.com/OData/odata.net/pull/1854/files#r470436103) comment could apply here too